File: app/src/main/java/com/project/scratchstudio/kith_andoid/Adapters/SearchAdapter.java
package com.project.scratchstudio.kith_andoid.Adapters;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import com.project.scratchstudio.kith_andoid.Holders.SearchHolder;
import com.project.scratchstudio.kith_andoid.R;
import com.project.scratchstudio.kith_andoid.Service.PicassoCircleTransformation;
import com.project.scratchstudio.kith_andoid.network.model.user.User;
import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class SearchAdapter extends RecyclerView.Adapter<SearchHolder> implements Filterable {

    public interface OnItemClickListener {
        void onItemClick(User item);
    }

    private List<User> searchList;
    private List<User> filteredData;

    private final OnItemClickListener listener;
    private Activity activity;

    public SearchAdapter(Activity activity, List<User> searchInfos, OnItemClickListener listener) {
        this.activity = activity;
        this.searchList = searchInfos;
        filteredData = searchInfos;
        this.listener = listener;
    }

    public void clearData() {
        searchList = null;
        filteredData = null;
    }

    @NonNull
    @Override
    public SearchHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View itemView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.list_item_layout, viewGroup, false);
        return new SearchHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull SearchHolder searchHolder, int i) {
        User searchInfo = filteredData.get(i);
        String name = searchInfo.firstName + " " + searchInfo.lastName;
        searchHolder.nName.setText(name);
        searchHolder.nPosition.setText(searchInfo.position);
        if (searchInfo.photo != null) {
            Picasso.with(activity).load(searchInfo.photo.replaceAll("@[0-9]*", ""))
                    .placeholder(com.project.scratchstudio.kith_andoid.R.mipmap.person)
                    .error(com.project.scratchstudio.kith_andoid.R.mipmap.person)
                    .transform(new PicassoCircleTransformation())
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(searchHolder.nPhoto);
        }

        searchHolder.bind(searchInfo, listener);
    }

    @Override
    public int getItemCount() {
        return filteredData.size();
    }

    public long getItemId(int position) {
        return position;
    }


    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                FilterResults results = new FilterResults();

                if (charSequence == null || charSequence.length() == 0) {
                    results.values = searchList;
                    results.count = searchList.size();
                } else {
                    String filter = String.valueOf(charSequence).toLowerCase();
                    ArrayList<User> filterResultsData = new ArrayList<>();

                    for (User data : searchList) {
                        String searchInfo = data.firstName + data.lastName + data.position + data.description;
                        boolean contain = true;
                        for (String filt : filter.split(" ")) {
                            if (!searchInfo.toLowerCase().contains(filt)) {
                                contain = false;
                            }
                        }
                        if (contain) {
                            filterResultsData.add(data);
                        }
                    }

                    results.values = filterResultsData;
                    results.count = filterResultsData.size();
                }

                return results;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                filteredData = (ArrayList<User>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }
}

File: app/src/main/java/com/project/scratchstudio/kith_andoid/adapters/SearchAdapter.kt
package com.project.scratchstudio.kith_andoid.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.project.scratchstudio.kith_andoid.R
import com.project.scratchstudio.kith_andoid.service.PicassoCircleTransformation
import com.project.scratchstudio.kith_andoid.holders.SearchHolder
import com.project.scratchstudio.kith_andoid.network.model.user.User
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.Picasso
import java.util.ArrayList

class SearchAdapter(private val activity: FragmentActivity, private val searchList: List<User>, private val listener: OnItemClickListener) : RecyclerView.Adapter<SearchHolder>(), Filterable {
    private var filteredData: List<User>? = null

    interface OnItemClickListener {
        fun onItemClick(item: User)
    }

    init {
        filteredData = searchList
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): SearchHolder {
        val itemView = LayoutInflater.from(viewGroup.context).inflate(R.layout.list_item_layout, viewGroup, false)
        return SearchHolder(itemView)
    }

    override fun onBindViewHolder(searchHolder: SearchHolder, i: Int) {
        val searchInfo = filteredData!![i]
        val name = searchInfo.firstName + " " + searchInfo.lastName
        searchHolder.nName.text = name
        searchHolder.nPosition.text = searchInfo.position
        if (searchInfo.photo != "") {
            Picasso.with(activity).load(searchInfo.photo.replace("@[0-9]*".toRegex(), ""))
                    .placeholder(R.mipmap.person)
                    .error(R.mipmap.person)
                    .transform(PicassoCircleTransformation())
                    .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                    .into(searchHolder.nPhoto)
        }

        searchHolder.bind(searchInfo, listener)
    }

    override fun getItemCount(): Int {
        return filteredData!!.size
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence?): FilterResults {
                val results = FilterResults()

                if (charSequence == null || charSequence.isEmpty()) {
                    results.values = searchList
                    results.count = searchList.size
                } else {
                    val filter = charSequence.toString().toLowerCase()
                    val filterResultsData = ArrayList<User>()

                    for (data in searchList) {
                        val searchInfo = data.firstName + data.lastName + data.position + data.description
                        var contain = true
                        for (filt in filter.split(" ".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()) {
                            if (!searchInfo.toLowerCase().contains(filt)) {
                                contain = false
                            }
                        }
                        if (contain) {
                            filterResultsData.add(data)
                        }
                    }

                    results.values = filterResultsData
                    results.count = filterResultsData.size
                }

                return results
            }

            override fun publishResults(charSequence: CharSequence, filterResults: Filter.FilterResults) {
                filteredData = filterResults.values as ArrayList<User>
                notifyDataSetChanged()
            }
        }
    }
}
